module/apms3session/session.go

```diff
@@ -0,0 +1,151 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apms3session // import "go.elastic.co/apm/module/apms3session"
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+	"go.elastic.co/apm/stacktrace"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+func init() {
+	stacktrace.RegisterLibraryPackage(
+		"github.com/aws/aws-sdk-go/aws/request",
+		"github.com/aws/aws-sdk-go/aws/session",
+	)
+}
+
+type handlers struct {
+	tracer *apm.Tracer
+}
+
+// Option sets options for tracing s3 requests.
+type Option func(*handlers)
+
+// WithTracer returns an Option which sets t as the tracer to use for tracing
+// s3 requests.
+func WithTracer(t *apm.Tracer) Option {
+	if t == nil {
+		panic("t == nil")
+	}
+	return func(h *handlers) {
+		h.tracer = t
+	}
+}
+
+func WrapSession(s *session.Session, opts ...Option) *session.Session {
+	h := &handlers{
+		tracer: apm.DefaultTracer,
+	}
+	for _, o := range opts {
+		o(h)
+	}
+
+	s.Handlers.Send.PushFrontNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/send",
+		Fn:   h.send,
+	})
+	s.Handlers.Complete.PushBackNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/complete",
+		Fn:   h.complete,
+	})
+
+	return s
+}
+
+const (
+	spanType    = "storage"
+	spanSubtype = "s3"
+)
+
+func (h *handlers) send(req *request.Request) {
+	if req.RetryCount > 0 {
+		return
+	}
+
+	var (
+		bucketName string
+
+		ctx    = req.Context()
+		region = *req.Config.Region
+	)
+
+	tx := apm.TransactionFromContext(ctx)
+	if tx == nil {
+		return
+	}
+
+	// TODO: All Params structs are required to have a `Bucket` attr. If
+	// there's a better way to get this, I would be overjoyed to avoid
+	// using runtime reflection.
+	params := reflect.ValueOf(req.Params).Elem()
+	if n, ok := params.FieldByName("Bucket").Interface().(*string); ok {
```

Comment:
Providing bucket name is required according to [the spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-aws.md#s3-simple-storage-service). All request objects (e.g. `s3.CreateBucketInput`) have a `Bucket` attribute, but they are stored as `interface{}` type on the request struct, hence the reflection. I'd gladly not do this, if there's something in the API that I missed that makes this easier to do.

Reply:
Each of the request objects has a `getBucket` method, but unexported :(
You could alternatively type switch on each request struct, but I don't think that's going to be very maintainable.

I think we should be able to extract the bucket from the request's Host. e.g. see the Host header in the examples in https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html. This is done in a default Build handler, so should be available by the time our Send handler is called.

Reply:
Dumping the request out (because I was getting `s3.us-west-2.amazonaws.com` in `req.URL.Host`, `req.Host` was empty string), I get:

```
PUT /BUCKET HTTP/1.1
Host: s3.us-west-2.amazonaws.com
User-Agent: aws-sdk-go/1.38.14 (go1.15.8; linux; amd64)
Content-Length: 153
Accept-Encoding: gzip
```
Which is odd because that isn't what the docs say. It could be a difference between v1 and v2 of the sdk, but it IS a bit worrisome. I tested it with the uploader for a second opinion and got a similar format:

```
PUT /BUCKET/some%20key HTTP/1.1
Host: s3.us-west-2.amazonaws.com
User-Agent: aws-sdk-go/1.38.14 (go1.15.8; linux; amd64) S3Manager
Content-Length: 16
Content-Md5: VwIh4BE1e7ENmaFxoY1EPA==
X-Amz-Content-Sha256: d49540b37fb41ad80dc4b6223720a91840587692ae752939d15ec53ae626e24b
Accept-Encoding: gzip
```
It's easy enough to take it from the path. I guess the latest docs in the link are showing the current API requests, which is probably what v2 of the sdk does.

Reply:
Ah hah:

```
	// Set this to `true` to force the request to use path-style addressing,
	// i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client
	// will use virtual hosted bucket addressing when possible
	// (`http://BUCKET.s3.amazonaws.com/KEY`).
	//
	// Note: This configuration option is specific to the Amazon S3 service.
	//
	// See http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html
	// for Amazon S3: Virtual Hosting of Buckets
	S3ForcePathStyle *bool
```

https://github.com/aws/aws-sdk-go/blob/main/aws/config.go#L118-L127

This is not currently being set, however, so it should be using the virtual hosted bucket subdomain, not the path-style addressing. I checked and the value is a nil pointer, but the documentation does say "By default, the S3 client will use virtual hosted bucket addressing when possible". Even when explicitly setting this value to `false`, it is still using the path style. So it appears that some requests will always behave like this.